### PR TITLE
(15350) Add fqdn_tag_rule resource

### PR DESCRIPTION
**Pull request by CyrusJavan**
- fqdn_domain_name resource
- Updated fqdn resouce domain_names to be Optional and Computed so that there is not an issue when using this new resource
- Updated docs on both fqdn and fqdn_domain_name resources

### CODE REVIEW

File: aviatrix/resource_aviatrix_fqdn.go [hunk 1]
```diff
@@ -62,6 +62,7 @@ func resourceAviatrixFQDN() *schema.Resource {
 			"domain_names": {
 				Type:        schema.TypeList,
 				Optional:    true,
+				Computed:    true,
```
Comment (arvindsree) [inline aviatrix/resource_aviatrix_fqdn.go]: Shouldn't make this computed. We should have a manage_domain_names flag similar to tgw resource

File: aviatrix/resource_aviatrix_fqdn_domain_name.go [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+	return &schema.Resource{
+		Create: resourceAviatrixFQDNDomainNameCreate,
+		Read:   resourceAviatrixFQDNDomainNameRead,
+		Delete: resourceAviatrixFQDNDomainNameDelete,
+		Importer: &schema.ResourceImporter{
+			State: schema.ImportStatePassthrough,
+		},
+
+		Schema: map[string]*schema.Schema{
+			"fqdn_tag": {
```
Comment (arvindsree) [inline aviatrix/resource_aviatrix_fqdn_domain_name.go]: rename to fqdn_tag_name

File: aviatrix/resource_aviatrix_fqdn_domain_name.go [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+				ForceNew:    true,
+				Description: "FQDN Filter Tag Name to attach this domain.",
+			},
+			"fqdn": {
+				Type:        schema.TypeString,
+				Required:    true,
+				ForceNew:    true,
+				Description: "FQDN.",
+			},
+			"proto": {
```
Comment (arvindsree) [inline aviatrix/resource_aviatrix_fqdn_domain_name.go]: rename to protocol

File: goaviatrix/fqdn.go [hunk 4]
```diff
@@ -483,3 +488,74 @@ func (c *Client) GetGwFilterTagList(fqdn *FQDN) (*FQDN, error) {
...
+}
+
+func (c *Client) GetFQDNDomainName(fqdn *FQDN) (*FQDN, error) {
+	foundFQDN, err := c.ListDomains(&FQDN{FQDNTag: fqdn.FQDNTag})
+	if err != nil {
+		return nil, fmt.Errorf("could not list fqdn domains: %v", err)
+	}
+	domain := fqdn.DomainList[0]
+	found := false
+	for _, d := range foundFQDN.DomainList {
```
Comment (arvindsree) [inline goaviatrix/fqdn.go:594]: Same. Better to have a new controller API to do this processing

File: aviatrix/resource_aviatrix_fqdn.go (hunk 1 above)
Comment (CyrusJavan) [reply aviatrix/resource_aviatrix_fqdn.go]: I'm curious, why is the manage_domain_names flag a better approach than Optional and Computed?

Comment (arvindsree) [reply aviatrix/resource_aviatrix_fqdn.go]: Making it computed will prevent managing domain_name any further on the fqdn resource. They will not be able to edit domain_names if they wish to do so.

File: goaviatrix/fqdn.go (hunk 4 above)
Comment (CyrusJavan) [reply goaviatrix/fqdn.go:594]: Opened mantis [16320](https://mantis.aviatrix.com/mantisbt/view.php?id=16320) proposing the new APIs for controller to do this processing.

File: aviatrix/resource_aviatrix_fqdn.go (hunk 1 above)
Comment (CyrusJavan) [reply aviatrix/resource_aviatrix_fqdn.go]: 👍

File: goaviatrix/fqdn.go (hunk 4 above)
Comment (arvindsree) [reply goaviatrix/fqdn.go:594]: Sounds good. We can keep the current code till the API is available

**Comment by CyrusJavan** [comment]
Renamed to `aviatrix_fqdn_tag_rule` per Sherry's request.

File: goaviatrix/fqdn.go [hunk 5]
```diff
@@ -557,3 +562,74 @@ func (c *Client) DisableFQDNPassThrough(gw *Gateway) error {
...
+	}{
+		Action:   action,
+		CID:      c.CID,
+		TagName:  fqdn.FQDNTag,
+		Policies: string(policies),
+	}, func(action string, reason string, ret bool) error {
+		if !ret {
+			// Tried to delete a rule that did not exist, we don't need to fail the apply.
+			if strings.Contains(reason, "the following rules were not found") {
+				return nil
```
Comment (arvindsree) [inline goaviatrix/fqdn.go]: We should not probably return ErrNotFound here and ignore the error for destroy code.

The acceptance usually checks for it to confirm if it has been destroyed

Comment (CyrusJavan) [reply goaviatrix/fqdn.go]: > ...should not probably return ErrNotFound here...

Do you mean we _should_ return ErrNotFound then just ignore it when calling the function?

Comment (arvindsree) [reply goaviatrix/fqdn.go]: Yeah..that's what I meant. This current approach works as well.

**Review by arvindsree — APPROVED** [review]

**Pull request merged**